File: 680_Valid_Palindrome_II.java
/*
https://leetcode.com/problems/valid-palindrome-ii/
Given a non-empty string s, you may delete at most one character. Judge whether you can make it a palindrome.

Set left and right indices to the ends of s. Iterate inwards. If characters do not match, either left or right must be
deleted so continue searching for a palindrome with no more deletions allowed.
Time - O(n)
Space - O(1)
*/

class Solution {
    public boolean validPalindrome(String s) {
        return helper(s, 0, s.length() - 1, false);
    }

    // deleted flags if a char has already been deleted
    private boolean helper(String s, int left, int right, boolean deleted) {

        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                if (deleted)        // already deleted a char
                    return false;
                return (helper(s, left, right - 1, true) || helper(s, left + 1, right, true)) ;
            }
            ++left;
            --right;
        }

        return true;
    }

}

File: python_1_to_1000/680_Valid_Palindrome_II.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/valid-palindrome-ii/
# Given a non-empty string s, you may delete at most one character. Judge whether you can make it a palindrome.

# Iterate over s from front and back. If chars do not match, check if the remainder after deleting either of the
# unmatched chars is a palindrome.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        n = len(s)
        i = 0

        while i < n // 2:

            if s[i] != s[n - 1 - i]:
                del_front = s[i + 1:n - i]
                del_back = s[i:n - 1 - i]
                return del_front == del_front[::-1] or del_back == del_back[::-1]

            i += 1

        return True